fastly/helpers.go

```diff
@@ -31,3 +33,22 @@ func NullString(v string) *string {
 	}
 	return &v
 }
+
+// ToSafeURL produces a safe (no path traversal, no unsafe characters) URL
+// from the path components passed in.
+func ToSafeURL(components ...string) string {
+	for idx, component := range components {
+		if component == ".." {
+			components[idx] = ""
+		} else {
+			components[idx] = url.PathEscape(component)
+		}
+	}
```

Comment:
Thanks! I'm about to push a new version of this function anyway, since I'm not comfortable doing an in-place modification of the slice argument. Also, the docs for `JoinPath` explicitly say that `../` components will be removed, so the loop doesn't need to check for those (and `PathEscape` does not modify them).

Reply:
I was incorrect; `JoinPath` does not remove `..` components, it resolves them, which isn't the behavior that we want. I've restored the test to ignore those components.

Reply:
Correct, they are “followed” if you imagine traversing the tree backwards. Important to check for them wherever untrusted input is expected.

I should have said so prior, but I looked at this pretty carefully for the review. Having tests that cover the expected edge cases is ideal, and comments on the function that it is critical to security and changes should be carefully vetted would be helpful.